plutus-core-interpreter/test/DynamicBuiltins/Definition.hs

```diff
@@ -0,0 +1,137 @@
+            TypeSchemeBuiltin (TypedBuiltinSized (SizeBound s) TypedBuiltinSizedInt) `TypeSchemeArrow`
+            TypeSchemeBuiltin (TypedBuiltinSized (SizeBound s) TypedBuiltinSizedInt)
+    fac n = product [1..n]
+
+dynamicFactorialDefinition :: DynamicBuiltinNameDefinition
+dynamicFactorialDefinition =
+    DynamicBuiltinNameDefinition dynamicFactorialName dynamicFactorialMeaning
+
+dynamicFactorial :: Term tyname name ()
+dynamicFactorial = dynamicBuiltinNameAsTerm dynamicFactorialName
+
+-- | Check that the dynamic factorial defined above computes to the same thing as
+-- a factorial defined in PLC itself.
+test_dynamicFactorial :: TestTree
+test_dynamicFactorial =
+    testCase "dynamicFactorial" $ do
+        let env = insertDynamicBuiltinNameDefinition dynamicFactorialDefinition mempty
+            lhs = typecheckEvaluateCek env $ applyFactorial dynamicFactorial 3 10
+            rhs = typecheckEvaluateCek mempty $ applyFactorial factorial 3 10
+        assertBool "type checks" $ isRight lhs
+        lhs @?= rhs
+
+dynamicConstName :: DynamicBuiltinName
+dynamicConstName = DynamicBuiltinName "const"
+
+dynamicConstMeaning :: DynamicBuiltinNameMeaning
+dynamicConstMeaning = DynamicBuiltinNameMeaning sch Prelude.const where
+    sch =
+        TypeSchemeAllType @"a" @0 Proxy $ \a ->
+        TypeSchemeAllType @"b" @1 Proxy $ \b ->
+            TypeSchemeBuiltin a `TypeSchemeArrow`
+            TypeSchemeBuiltin b `TypeSchemeArrow`
+            TypeSchemeBuiltin a
+
+dynamicConstDefinition :: DynamicBuiltinNameDefinition
+dynamicConstDefinition =
+    DynamicBuiltinNameDefinition dynamicConstName dynamicConstMeaning
+
+dynamicConst :: Term tyname name ()
+dynamicConst = dynamicBuiltinNameAsTerm dynamicConstName
+
+-- | Check that the dynamic const defined above computes to the same thing as
+-- a const defined in PLC itself.
+test_dynamicConst :: TestTree
+test_dynamicConst =
+    testProperty "dynamicConst" . property $ do
+        c <- forAll Gen.unicode
+        b <- forAll Gen.bool
+        let tC = fromMaybe (Prelude.error "Can't make a char") $ makeDynamicBuiltin c
+            tB = fromMaybe (Prelude.error "Can't make a bool") $ makeDynamicBuiltin b
+            char = toTypeEncoding @Char Proxy
+            runConst con = mkIterApp () (mkIterInst () con [char, bool]) [tC, tB]
+            env = insertDynamicBuiltinNameDefinition dynamicConstDefinition mempty
+            lhs = typecheckReadDynamicBuiltinCek env $ runConst dynamicConst
+            rhs = typecheckReadDynamicBuiltinCek mempty $ runConst Plc.const
+        lhs === Right (Right (EvaluationSuccess c))
+        lhs === rhs
+
+dynamicReverseName :: DynamicBuiltinName
+dynamicReverseName = DynamicBuiltinName "reverse"
+
+dynamicReverseMeaning :: DynamicBuiltinNameMeaning
+dynamicReverseMeaning = DynamicBuiltinNameMeaning sch (PlcList . Prelude.reverse . unPlcList) where
+    sch =
+        TypeSchemeAllType @"a" @0 Proxy $ \(_ :: TypedBuiltin size a) ->
```

Comment:
Hang on, shouldn't the argument to the lambda correspond to the bound type variable, i.e. the type of the elements of the list, not a size?

Reply:
It does correspond to the type of the elements of the list. The `size` variable is bound only because we can't really omit it without enabling some extension. I can alias `TypedBuiltin size` to `bound` the same way I alias `QuotedTerm text uniq` to `a`, so that you can write

```
TypeSchemeAllType @"a" @0 Proxy $ \(_ :: bound a) ->
```

but I'm not sure if I should do that, because in the `const` test above we use the fact that the bound term-level variable is a `TypedBuiltin`, so another alias may look misleading (even though it does not break anything).